x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/InferModelAction.java

```diff
@@ -0,0 +1,188 @@
+            this.cacheModel = true;
+            this.topClasses = topClasses;
+        }
+
+        public Request(String modelId, long modelVersion, Map<String, Object> objectToInfer, Integer topClasses) {
+            this(modelId,
+                modelVersion,
+                objectToInfer == null ? null : Arrays.asList(objectToInfer),
+                topClasses);
+        }
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
+            this.modelId = in.readString();
+            this.modelVersion = in.readVLong();
+            this.objectsToInfer = Collections.unmodifiableList(in.readList(StreamInput::readMap));
+            this.topClasses = in.readOptionalInt();
+            this.cacheModel = in.readBoolean();
+        }
+
+        public String getModelId() {
+            return modelId;
+        }
+
+        public long getModelVersion() {
+            return modelVersion;
+        }
+
+        public List<Map<String, Object>> getObjectsToInfer() {
+            return objectsToInfer;
+        }
+
+        public boolean isCacheModel() {
+            return cacheModel;
+        }
+
+        public Integer getTopClasses() {
+            return topClasses;
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            return null;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            out.writeString(modelId);
+            out.writeVLong(modelVersion);
+            out.writeCollection(objectsToInfer, StreamOutput::writeMap);
+            out.writeOptionalInt(topClasses);
+            out.writeBoolean(cacheModel);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            InferModelAction.Request that = (InferModelAction.Request) o;
+            return Objects.equals(modelId, that.modelId)
+                && Objects.equals(modelVersion, that.modelVersion)
+                && Objects.equals(topClasses, that.topClasses)
+                && Objects.equals(cacheModel, that.cacheModel)
+                && Objects.equals(objectsToInfer, that.objectsToInfer);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(modelId, modelVersion, objectsToInfer, topClasses, cacheModel);
+        }
+
+    }
+
+    public static class RequestBuilder extends ActionRequestBuilder<Request, Response> {
+        public RequestBuilder(ElasticsearchClient client, Request request) {
+            super(client, INSTANCE, request);
+        }
+    }
+
+    public static class Response extends ActionResponse {
+
+        private final List<InferenceResults<?>> inferenceResponse;
+        private final String resultsType;
+
+        public Response(List<InferenceResults<?>> inferenceResponse, String resultsType) {
```

Comment:
shall we call `inferenceResponse` -> `inferenceResults`? Also for the member variable.